Toast rule (#228)
avito-tech/avito-android: docs/content/docs/test/Toast.md

---
title: Testing toasts
type: docs
---

> # Testing toasts

Reviewer: Toast testing?

Author: original one looks better to me

[android.widget.Toast](https://developer.android.com/reference/android/widget/Toast)

{{< hint warning>}}
Testing toasts reliably is a hard task because some other toast can show up and yours will be queued. 

> We hack app with proxy object to mitigate this.

Reviewer: ```suggestion
Testing toasts is a hard because other toasts can show up and yours will be queued. 
We hack app with proxy object to mitigate this.
```

{{< /hint >}}

{{< hint warning>}}
It is also recommended to use [snackbar](https://developer.android.com/reference/com/google/android/material/snackbar/Snackbar), 
or your custom way to display brief unintrusive messages instead. So you should consider not testing it at all.
{{< /hint >}}

## Proxy toast

Use provided extension methods to be able to spy on `showToast` functions.

```kotlin
import com.avito.android.util.showToast

showToast("I'am a toast!")
```

```kotlin

> implementation("com.avito.android:proxy-toast:$version")

Reviewer: Do we need describe all deps?

Author: what deps? it's enough for this library to run

Reviewer: I understand nothing.
You created two modules `proxy-toast` and `toast-rule`. You added them to test-app dependencies.
Is `proxy-toast` enough for what without toast-rule?
You described usages of ToastRule but didn't add description how to add it dependency. Comparing with `proxy-toast` why didn't you add dependency description?

Author: yep, i missed that, thanks

```

## ToastRule

> ToastRule changes proxy toast's singleton before the test to store `showToast` invocations.

Reviewer: Change abstract `changes`
```suggestion
ToastRule clears shown toast history that stores in proxy toast singleton before the test.
```

Author: i don't understand the phrase now

Reviewer: I've updated

Author: @eugene-krivobokov , @int02h, what do you think? which variant is better?

Reviewer 2: I'd prefer to avoid these implementation details here.
For me, the phrase "Use provided extension methods to be able to spy ... " was enough.  I don't need to know about singleton and history of invocations.

Reviewer 3: Maybe "Before the test ToastRule clears history of shown toasts stored in proxy toast singleton."?

Author: it is not necessary to clear collection before test as we run all of them in isolation right now;
i don't want to focus on it here.
All i wanted to say that Rule changes singleton value to it's own Proxy object and keeps tracking of invocations. 
I think my original explanation good enough

Reviewer 2: >  Rule changes singleton value to it's own Proxy object and keeps tracking of invocations.

The more I read this the more I think that this is excessive implementation details.

Reviewer: What is a value of toatrule info? Implementation details that teach how to use rule correctly give a value.

---
title: Testing toasts
type: docs
---

# Testing toasts

[android.widget.Toast](https://developer.android.com/reference/android/widget/Toast)

{{< hint warning>}}
Testing toasts is a hard because other toasts can show up and yours will be queued. 
We hack app with proxy object to mitigate this.
{{< /hint >}}

{{< hint warning>}}
It is also recommended to use [snackbar](https://developer.android.com/reference/com/google/android/material/snackbar/Snackbar), 
or your custom way to display brief unintrusive messages instead. So you should consider not testing it at all.
{{< /hint >}}

## Proxy toast

Use provided extension methods to be able to spy on `showToast` functions.

```kotlin
import com.avito.android.util.showToast

showToast("I'am a toast!")
```

```kotlin
implementation("com.avito.android:proxy-toast:$version")
```

## ToastRule

```kotlin
@get:Rule
val toastRule = ToastRule()

fun test() {
    toastRule.checks.toastDisplayedWithText("I'am a toast!")
}
```

```kotlin
androidTestImplementation("com.avito.android:toast-rule:$version")

> ```

Reviewer: It's look like abstract bull shit.
Point of this page isn't clear.
It's not clear that `showToast` must be used in production instead of standart functions.
It's not clear that Rule useless without `showToast`
Info about SnackBar is too large compare with other content.
Structure i see:
- You can't test toast
- We help you to test toast
- You can test toast that shows only by our `showToast` in production
- You can use our rule
- We recommend to use snackbar

Reviewer 2: > It's look like abstract bull shit.

I see your reaction and share some worries. But please, try to find other words. The problem is, it can be interpreted as aggression and doesn't provide information about exact problems.
https://avito-tech.github.io/avito-android/docs/contributing/codereview/#best-practices

Apart from this, I agree with the next proposals.

Reviewer: Sorry @dsvoronin . I want to say that text is hard to read. I will work on my messages

Reviewer: Do i need to create pr with proposals?
